File: 1568.py
class Solution:
    def minDays(self, grid: list[list[int]]) -> int:
        n = len(grid)
        m = len(grid[0])
        def count_islands(grid):
            def dfs(x, y):
                if x < 0 or x >= n or y < 0 or y >= m or grid[x][y] != 1:
                    return
                grid[x][y] = 2
                dfs(x + 1, y)
                dfs(x - 1, y)
                dfs(x, y + 1)
                dfs(x, y - 1)
                
            count = 0
            for i in range(n):
                for j in range(m):
                    if grid[i][j] == 1:
                        dfs(i, j)
                        count += 1
            for i in range(n):
                for j in range(m):
                    if grid[i][j] == 2:
                        grid[i][j] = 1
            return count
        
        if count_islands(grid) != 1:
            return 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    grid[i][j] = 0
                    if count_islands(grid) != 1:
                        return 1
                    grid[i][j] = 1
        return 2
    
if __name__ == '__main__':
    [n, m] = list(map(int, input().split()))
    grid = [list(map(int, input().split())) for _ in range(n)]
    print(Solution().minDays(grid))

File: cpp/1568.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution {
public:
    int minDays(vector<vector<int>>& grid){
        n = grid.size();
        m = grid[0].size();
        if(countIslands(grid) != 1) // more than 1 island
            return 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 1){
                    grid[i][j] = 0; // remove the island
                    if(countIslands(grid) != 1) // more than 1 island
                        return 1;
                    grid[i][j] = 1; // restore the island
                }
            }
        }
        // we can't remove any single island to make it disconnected
        return 2; // we can isolate an island by removing 2 of its adjacent islands
    }
private:
    int m, n;
    int countIslands(vector<vector<int>>& grid){
        int count = 0;
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 1){
                    dfs(grid, i, j);
                    count++;
                }
            }
        }
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == 2) grid[i][j] = 1;
            }
        }
        return count;
    }
    void dfs(vector<vector<int>>& grid, int i, int j){
        if(i < 0 || j < 0 || i >= n || j >= m || grid[i][j] != 1) return;
        grid[i][j] = 2;
        dfs(grid, i+1, j);
        dfs(grid, i-1, j);
        dfs(grid, i, j+1);
        dfs(grid, i, j-1);
    }
};

int main(){
    int n, m;
    cin >> n >> m;
    vector<vector<int>> grid(n, vector<int>(m));
    for(int i = 0; i < n; i++)
        for(int j = 0; j < m; j++)
            cin >> grid[i][j];
    Solution s;
    cout << s.minDays(grid) << endl;
    return 0;
}